keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,669 @@
+
+In addition to the new catalog kind, `kustomize build` will accept a repeatable flag `--trusted-catalog=""`. The values passed to this flag should match the values declared within the catalog. When present, this flag instructs `kustomize build` to automatically fetch and execute functions that are defined by the catalog and referenced within the Kustomization, Component or Composition, if needed. When a resource is processed by `kustomize build` and a catalog is referenced but not specified using the `--trusted-catalog=""` flag, an error will occur. 
+
+To support the user experience of using catalogs, we will also add a `kustomize view catalog` command that will display catalog configuration, including publisher information retrieved from the catalog itself, in order to support the trust determination process.
+
+Kustomize can at a later date provide a built in Catalog for supporting official extensions, published to a well publicized endpoint. This catalog will _NOT_ require the user to explicitly trust it. Users can provide the `apiVersion` and `kind` of the official extensions in kustomize resource and these will be resolved by the official catalog.
+
+In addition to container based functions, the `Catalog` will support discovery of Starlark and Exec based functions, via an HTTP(s), Git, or OCI reference as illustrated below: 
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-functions"
+spec: 
+  providers: 
+  - apiVersion: example.com/v1
+    kind: GroovyApplication
+    description: "A Kustomize function that can handle groovy apps"
+    provider:  
+      starlark: https://example.co/module_providers/starlark-func:v1.0.0
+```
+
+This concept can be extended later to support additional function packaging, but is out of scope for the current proposal.
+
+When HTTP(s) references are used, the HTTP(s) endpoint must support anonymous access for reading resources. Resources will be expected to be stored as a single file, such as `catalog.yaml`. Git or OCI references can be authenticated or anonymous, and will use appropriate configuration from the users file-system. 
+
+ 
+### User Stories (Optional)
+
+#### Story 1
+
+As a platform developer at enterprise company Example Co, I want to publish a catalog of Kustomize functions that represent custom capabilities important for our internal Kubernetes platform. I have built and published several Kustomize functions, packaged as Docker images, to our internal Docker registry: `docker.example.co`.
+
+To do this, I build a new `Catalog` API resource: 
+
+```yaml
+# catalog.yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-functions"
+spec: 
+  providers: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize function that can handle Java apps"
+    provider: 
+      container: 
+        image: docker.example.co/functions/java:v1.0.0
+  - apiVersion: example.com/v1
```

Comment:
Suppose we initially have `docker.example.co/functions/logger:v1.0.0` that only supports `functionConfig` with `apiVersion` = `example.com/v1alpha1` and `kind`=`Logger`. 
We decide to promote the functionConfig from `v1alpha1` to `v1beta1` and we want to new function `docker.example.co/functions/logger:v1.1.0` to support both `apiVerison`s.

As a platform developer, usually I cannot force all of my users in my organization to update the `apiVersion` from v1alpha1 to v1beta1. I will need to provide both `apiVersion`s in my catalog.

How should we present it in the `Catalog` object? Will it look like the following? It's too verbose and I'm repeating myself.

```yaml
apiVersion: kustomize.io/v1
kind: Catalog
metadata: 
  name: "example-co-functions"
spec: 
  providers: 
  - apiVersion: example.com/v1alpha1
    kind: Logger
    description: "A Kustomize function that adds our bespoke logging"
    provider: 
      container: 
        image: docker.example.co/functions/logger:v1.1.0
  - apiVersion: example.com/v1beta1
    kind: Logger
    description: "A Kustomize function that adds our bespoke logging"
    provider: 
      container: 
        image: docker.example.co/functions/logger:v1.1.0
```

Reply:
We are doing that currently in our internal proof of concept:

Reply:
What do you think about something like the following?

```yaml
apiVersion: kustomize.io/v1
kind: Catalog
metadata: 
  name: "example-co-functions"
spec: 
  providers: 
  - apiVersion: example.com/v1beta1    # primary apiVersion
    kind: Logger                       # primary kind
    additionalAPIVersionKinds:
      - apiVersion: example.com/v1alpha1
        kind: Logger
        deprecated: true
      - apiVersion: example.com/v1alpha1
        kind: LegacyLogger
        deprecated: true
    description: "A Kustomize function that adds our bespoke logging"
    provider: 
      container: 
        image: docker.example.co/functions/logger:v1.1.0
```

Reply:
that sounds good to me.

Reply:
I like that the new design clearly shows the relationship between multiple API versions. Two questions:
- In situations where a given provider supports multiple unrelated GVKs, we still want separate top-level entries, right? Might it be reasonable to not allow the Kind to differ in the "additional" subentries?
- Consider the case where there are multiple providers (e.g. an exec and a container). This layout implies they will all support all listed GVs and be updated in lock step. Is that a reasonable requirement? E.g. what if you introduce an exec option at v1, and do not want to implement v1alpha1 support. Will we end up inlining all the top-level fields under "additional" as overrides?

Reply:
> In situations where a given provider supports multiple unrelated GVKs, we still want separate top-level entries, right? Might it be reasonable to not allow the Kind to differ in the "additional" subentries?

If a provider supports multiple GVKs that are completely unrelated, ideally they should be separate top-level entries. For example, if I have apiVersions `example.com/v1` and `example.com/v2`, they should be listed as separate top-level entries, even though one single KRM function may support both of them.
`v1.ConfigMap` as functionConfig might be something special here. I think we all agreed that the primary GVK should always be a strong typed client-side CRD, and`v1.ConfigMap` should never be used as the primary GVK. But IMO it'd be valuable to support `ConfigMap` as a secondary GVK.
IMO we should allow some flexibility but also provide a best practice doc about how to organize them.
@KnVerey @jeremyrickard WDYT?

> Consider the case where there are multiple providers (e.g. an exec and a container). This layout implies they will all support all listed GVs and be updated in lock step. Is that a reasonable requirement? E.g. what if you introduce an exec option at v1, and do not want to implement v1alpha1 support. Will we end up inlining all the top-level fields under "additional" as overrides?

When a user have a exec-based KRM function, it's fairly easy to put it in a container. Then they will support the same GVKs. But this practice doesn't apply to starlark provider. Same as above, we should provide a best practice doc about how to organize them.

@KnVerey @jeremyrickard  Can someone please give me a pointer to the description of the `lock step`?

We will not end up inlining all the top-level fields under "additional" as overrides. The users must use different top-level entries in at least the following cases:
- Major KRM version bump. e.g. from `example.com/v1*` to `example.com/v2*`.
- The provider don't support all of these GVKs. e.g. I have `docker.example.co/functions/logger:v1.0.0` that don't support `example.com/v1beta1`. Our catalog may looks like:
```
apiVersion: kustomize.io/v1
kind: Catalog
metadata: 
  name: "example-co-functions"
spec: 
  providers: 
  - apiVersion: example.com/v1beta1
    kind: Logger
    additionalAPIVersionKinds:
      - apiVersion: example.com/v1alpha1
        kind: Logger
        deprecated: true
      - apiVersion: example.com/v1alpha1
        kind: LegacyLogger
        deprecated: true
    description: "A Kustomize function that adds our bespoke logging"
    provider: 
      container: 
        image: docker.example.co/functions/logger:v1.1.0
  - apiVersion: example.com/v1alpha1
    kind: Logger
    additionalAPIVersionKinds:
      - apiVersion: example.com/v1alpha1
        kind: LegacyLogger
        deprecated: true
    description: "A Kustomize function that adds our bespoke logging"
    provider: 
      container: 
        image: docker.example.co/functions/logger:v1.0.0
```
As descrived in this KEP, the order matters in the `spec.providers` list. If a user specify `apiVersion: example.com/v1alpha1` and `kind: Logger`, the provider `docker.example.co/functions/logger:v1.1.0` is prefered over `docker.example.co/functions/logger:v1.0.0`.
Thoughts?